plugins/provisioners/shell/provisioner.rb

```diff
@@ -135,24 +137,22 @@ def provision_winssh(args)
           # Upload the script to the machine
           @machine.communicate.tap do |comm|
             env = config.env.map{|k,v| comm.generate_environment_export(k, v)}.join(';')
-            shell = comm.machine_config_ssh.shell
-            if File.extname(upload_path).empty?
-              remote_ext = File.extname(path)[1..-1]
-              if !remote_ext
-                remote_ext = @machine.config.winssh.shell == "cmd" ? "bat" : "ps1"
-              end
-              upload_path << ".#{remote_ext}"
+            remote_ext = File.extname(path)
+            if remote_ext == ""
+              remote_ext = @machine.config.winssh.shell == "cmd" ? ".bat" : ".ps1"
             end
+            remote_path = add_extension(upload_path, remote_ext)
```

Comment:
just overwrite `upload_path` here and then keep the name consistent below.

Reply:
The `upload_path` here refers to this function https://github.com/hashicorp/vagrant/blob/master/plugins/provisioners/shell/provisioner.rb#L41. So, setting `upload_path = add_extension(upload_path, remote_ext)` causes some strange behaviors. I think something like `remote_path = add_extension(upload_path(), remote_ext)` is maybe less ambiguous. I'm not sure what the ruby way is here?

Reply:
Completely valid assertion and my mistake :slightly_smiling_face: